File: snippets/lesson-1-observable.ts
import { Observable, Subscription } from 'rxjs';

const numberStream$: Observable<number> = new Observable<number>((subscriber) => {
  subscriber.next(1);
  subscriber.next(2);
  subscriber.next(3);
  subscriber.complete();
});

const subscription: Subscription = numberStream$.subscribe({
  next: (value: number) => console.log(`Received: ${value}`),
  error: (err: Error) => console.error(`Error: ${err.message}`),
  complete: () => console.log('Stream completed!'),
});

subscription.unsubscribe();

File: snippets/lesson-1-observable.cs
using System;
using System.Reactive.Disposables;
using System.Reactive.Linq;

IObservable<int> numberStream = Observable.Create<int>(observer =>
{
    observer.OnNext(1);
    observer.OnNext(2);
    observer.OnNext(3);
    observer.OnCompleted();
    return Disposable.Empty;
});

IDisposable subscription = numberStream.Subscribe(
    onNext: (int value) => Console.WriteLine($"Received: {value}"),
    onError: (Exception err) => Console.WriteLine($"Error: {err.Message}"),
    onCompleted: () => Console.WriteLine("Stream completed!")
);

subscription.Dispose();